File: data/backtracking/n-queens.py
def solveNQueens(n: int) -> list[list[str]]:
    """
    N-Queens Backtracking Algorithm
    Time Complexity: O(N!)
    Space Complexity: O(N^2)
    """
    res = []
    cols = set()
    posDiag = set() # r + c
    negDiag = set() # r - c
    board = [["."] * n for _ in range(n)]

    def backtrack(r):
        if r == n:
            res.append(["".join(row) for row in board])
            return

        for c in range(n):
            if c in cols or (r + c) in posDiag or (r - c) in negDiag:
                continue

            cols.add(c)
            posDiag.add(r + c)
            negDiag.add(r - c)
            board[r][c] = "Q"

            backtrack(r + 1)

            cols.remove(c)
            posDiag.remove(r + c)
            negDiag.remove(r - c)
            board[r][c] = "."

    backtrack(0)
    return res

File: data/backtracking/n-queens.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

class SolutionNQueens {
    vector<vector<string>> res;
    unordered_set<int> cols, posDiag, negDiag;

    void backtrack(int r, int n, vector<string>& board) {
        if (r == n) {
            res.push_back(board);
            return;
        }

        for (int c = 0; c < n; ++c) {
            if (cols.count(c) || posDiag.count(r + c) || negDiag.count(r - c)) {
                continue;
            }

            cols.insert(c);
            posDiag.insert(r + c);
            negDiag.insert(r - c);
            board[r][c] = 'Q';

            backtrack(r + 1, n, board);

            cols.erase(c);
            posDiag.erase(r + c);
            negDiag.erase(r - c);
            board[r][c] = '.';
        }
    }

public:
    vector<vector<string>> solveNQueens(int n) {
        vector<string> board(n, string(n, '.'));
        backtrack(0, n, board);
        return res;
    }
};
